zulipterminal/ui_tools/buttons.py

```diff
@@ -261,14 +262,52 @@ def __init__(self, stream_id: int, topic: str,
                          count=count)
 
         if controller.model.is_muted_topic(self.stream_id, self.topic_name):
-            self.mark_muted()
+            self.update_widget(('muted',  MUTE_MARKER), 'muted')
+            # TODO: Handle event-based approach for topic-muting.
+
+    def set_button_counts(self, unmuted_count: int) -> None:
+        self.update_count(unmuted_count)
+        self.model.unread_counts['streams'][self.stream_id] += unmuted_count
+        stream_button = self.view.stream_id_to_button[self.stream_id]
+        stream_button.update_count(
+            self.model.unread_counts['streams'][self.stream_id])
+        self.model.unread_counts['all_msg'] += unmuted_count
+        self.view.home_button.update_count(
+            self.model.unread_counts['all_msg'])
 
     def mark_muted(self) -> None:
+        """
+        We do not decrease the count of topic button (or
+        unread_counts['unread_topics']) but only mark it as M since
+        we would require the correct count while unmuting.
+        However, we decrease the all msg count and the topic's stream
+        count by the unread counts.
+        """
         self.update_widget(('muted',  MUTE_MARKER), 'muted')
-    # TODO: Handle event-based approach for topic-muting.
+        if self.stream_id in self.model.unread_counts['streams']:
+            self.model.unread_counts['streams'][self.stream_id] -= self.count
+            stream_button = self.view.stream_id_to_button[self.stream_id]
+            stream_button.update_count(
+                self.model.unread_counts['streams'][self.stream_id])
+        self.model.unread_counts['unread_muted_topics'][(
+                                self.stream_id, self.topic_name)] = self.count
+        self.model.unread_counts['unread_topics'].pop(
+                                    (self.stream_id, self.topic_name), None)
+        self.model.unread_counts['all_msg'] -= self.count
```

Comment:
This needs an edge case handling. (See my review bullet)

Reply:
I haven't checked other updates, though I could still reproduce this behavior (i.e., subtracting unreads in the case of a muted stream)